src/orange/detail/BIHBuilder.cc

```diff
@@ -150,64 +149,67 @@ void BIHBuilder::construct_tree(VecIndices const& indices,
 
 //---------------------------------------------------------------------------//
 /*!
- * Seperate nodes into inner and leaf vectors and renumber accordingly.
+ * Partition nodes into inner and leaf vectors and renumber accordingly.
  */
 BIHBuilder::ArrangedNodes BIHBuilder::arrange_nodes(VecNodes nodes) const
 {
     VecInnerNodes inner_nodes;
     VecLeafNodes leaf_nodes;
 
     std::vector<bool> is_leaf;
-    std::vector<BIHNodeId> new_ids;
-
-    auto visit_node
-        = Overload{[&](BIHInnerNode const& node) {
-                       new_ids.push_back(BIHNodeId(inner_nodes.size()));
-                       inner_nodes.push_back(node);
-                       is_leaf.push_back(false);
-                   },
-                   [&](BIHLeafNode const& node) {
-                       new_ids.push_back(BIHNodeId(leaf_nodes.size()));
-                       leaf_nodes.push_back(node);
-                       is_leaf.push_back(true);
-                   }};
-
+    std::vector<std::size_t> new_idx;
+
+    is_leaf.reserve(nodes.size());
+    new_idx.reserve(nodes.size());
+
+    auto insert_node = Overload{[&](BIHInnerNode const& node) {
+                                    new_idx.push_back(inner_nodes.size());
+                                    inner_nodes.push_back(node);
+                                    is_leaf.push_back(false);
+                                },
+                                [&](BIHLeafNode const& node) {
+                                    new_idx.push_back(leaf_nodes.size());
+                                    leaf_nodes.push_back(node);
+                                    is_leaf.push_back(true);
+                                }};
     for (auto const& node : nodes)
     {
-        std::visit(visit_node, node);
+        std::visit(insert_node, node);
     }
 
-    size_type offset = inner_nodes.size();
-
+    // Transform "leaf ID" to "node ID"
+    auto offset = inner_nodes.size();
     for (auto i : range(nodes.size()))
     {
         if (is_leaf[i])
         {
-            new_ids[i] = new_ids[i] + offset;
+            new_idx[i] += offset;
         }
     }
 
+    // Remap IDs. "parent" will only be undefined for the root node.
+    auto remapped_id = [&new_idx](BIHNodeId old) {
+        CELER_EXPECT(old < new_idx.size());
+        return BIHNodeId{static_cast<size_type>(new_idx[old.unchecked_get()])};
+    };
+
     for (auto& inner_node : inner_nodes)
     {
-        for (auto edge : range(BIHInnerNode::Edge::size_))
+        for (auto& bp : inner_node.bounding_planes)
         {
-            inner_node.bounding_planes[edge].child
-                = new_ids[inner_node.bounding_planes[edge].child.unchecked_get()];
+            bp.child = remapped_id(bp.child);
         }
-
-        // Handle root node
```

Comment:
It wasn't at all clear to me that the comment was describing the `if (false)` case: the remapping needed to be done in all cases *except* fro the root node. Putting the comment inside the `if` like `// Not the root node: update its parent ID`